ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/utils/IncompleteInputFinder.java

```diff
@@ -47,17 +71,40 @@
 
     @Override
     public Boolean transform(SimpleNameReferenceNode node) {
-        return node.name().isMissing();
+        // Check node diagnostics since parseAsExpression("foreach Employee e in t") recovered as
+        // SimpleNameReferenceNode
+        return node.name().isMissing() || node.hasDiagnostics();
+    }
+
+    @Override
+    public Boolean transform(BuiltinSimpleNameReferenceNode node) {
+        // Check node diagnostics since parseAsExpression("foreach float x in v") recovered as
+        // BuiltinSimpleNameReferenceNode
+        return node.name().isMissing() || node.hasDiagnostics();
     }
 
     @Override
     public Boolean transform(BinaryExpressionNode node) {
         return node.rhsExpr().apply(this);
     }
 
+    @Override
+    public Boolean transform(FunctionDefinitionNode node) {
+        if (node.functionKeyword().isMissing()) {
+            return true;
+        }
```

Comment:
What if we combine this with line 97? Same applies for the other transform methods as well

Reply:
Updated in 7d25e1dea188151fc8c5f84a1bf9156ce22ebc81. Updated other possible cases as well.